include/mbedtls/lms.h

```diff
@@ -0,0 +1,254 @@
+/**
+ * \file lms.h
+ *
+ * \brief This file provides an API for the LMS post-quantum-safe stateful-hash
+ *        public-key signature scheme.
+ */
+/*
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+#ifndef MBEDTLS_LMS_H
+#define MBEDTLS_LMS_H
+
+#include <stdint.h>
+#include <stddef.h>
+
+#include "mbedtls/private_access.h"
+#include "mbedtls/lmots.h"
+
+#define MBEDTLS_ERR_LMS_BAD_INPUT_DATA   -0x0011 /**< Bad data has been input to an LMS function */
+#define MBEDTLS_ERR_LMS_OUT_OF_PRIV_KEYS -0x0013 /**< Specified LMS key has utilised all of its private keys */
+#define MBEDTLS_ERR_LMS_VERIFY_FAILED    -0x0015 /**< LMS signature verification failed */
+#define MBEDTLS_ERR_LMS_ALLOC_FAILED     -0x0017 /**< LMS failed to allocate space for a private key */
+
+#define MBEDTLS_LMS_TYPE_LEN        (4)
+#define MBEDTLS_LMS_H_TREE_HEIGHT       (10)
+#define MBEDTLS_LMS_M_NODE_BYTES        (32)
+
+#define MBEDTLS_LMS_SIG_LEN (MBEDTLS_LMOTS_Q_LEAF_ID_LEN + MBEDTLS_LMOTS_SIG_LEN + \
+                             MBEDTLS_LMS_TYPE_LEN + MBEDTLS_LMS_H_TREE_HEIGHT * MBEDTLS_LMS_M_NODE_BYTES)
+
+#define MBEDTLS_LMS_PUBKEY_LEN (MBEDTLS_LMS_TYPE_LEN + MBEDTLS_LMOTS_TYPE_LEN + \
+                                MBEDTLS_LMOTS_I_KEY_ID_LEN + MBEDTLS_LMS_M_NODE_BYTES)
+
+#define MBEDTLS_LMS_SIG_Q_LEAF_ID_OFFSET    (0)
+#define MBEDTLS_LMS_SIG_OTS_SIG_OFFSET      (MBEDTLS_LMS_SIG_Q_LEAF_ID_OFFSET + MBEDTLS_LMOTS_Q_LEAF_ID_LEN)
+#define MBEDTLS_LMS_SIG_TYPE_OFFSET         (MBEDTLS_LMS_SIG_OTS_SIG_OFFSET   + MBEDTLS_LMOTS_SIG_LEN)
+#define MBEDTLS_LMS_SIG_PATH_OFFSET         (MBEDTLS_LMS_SIG_TYPE_OFFSET      + MBEDTLS_LMS_TYPE_LEN)
+
+#define MBEDTLS_LMS_PUBKEY_TYPE_OFFSET      (0)
+#define MBEDTLS_LMS_PUBKEY_OTSTYPE_OFFSET   (MBEDTLS_LMS_PUBKEY_TYPE_OFFSET     + MBEDTLS_LMS_TYPE_LEN)
+#define MBEDTLS_LMS_PUBKEY_I_KEY_ID_OFFSET  (MBEDTLS_LMS_PUBKEY_OTSTYPE_OFFSET  + MBEDTLS_LMOTS_TYPE_LEN)
+#define MBEDTLS_LMS_PUBKEY_ROOT_NODE_OFFSET (MBEDTLS_LMS_PUBKEY_I_KEY_ID_OFFSET + MBEDTLS_LMOTS_I_KEY_ID_LEN)
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum {
+    MBEDTLS_LMS_SHA256_M32_H10 = 0x6,
+} mbedtls_lms_algorithm_type_t;
+
+
+typedef struct {
+    unsigned MBEDTLS_PRIVATE(have_privkey);
```

Comment:
Use of `bool` seems patch in mbedtls (I'm not sure if that's a coding guideline thing), and I suspect most of the time we'd end up not saving space because of the way struct packing usually works, but I'm happy to change if you think it'd be benificial.

Reply:
We have some internal documentation (should be public really, but we haven't got around to moving it yet, sorry) about usage of C99 features (which `bool` is), and it says that `stdbool.h` is still a no, because MSVC 2017 doesn't have it (and we still support that version, and even back to MSVC 2013 IIRC).

Using `uint8_t` sounds good to. As you point out, depending on struct padding it may not make a difference, but OTOH if there are several of them we can group them together and then it does make a difference.

On a related note: at some point we used bitfields (`unsigned foo : 1;`) but we moved away from them because they tend to result in larger code than `uint8_t`.

Finally: please mention in the field's documentation that it's actually a bool, though.

Reply:
@mpg `stdbool.h` was actually added [in VS 2013](https://devblogs.microsoft.com/cppblog/c99-library-support-in-visual-studio-2013/), which is the earliest version we support. But we haven't started using it (except in Everest which comes with its own platform support), I think because it has little benefit: a `bool` takes as much space in memory as a `char` anyway, so we tend to stick with `uint8_t` in structures and `int` for boolean local variables and return values. And `uint8_t` fields have the advantage that it's not an ABI/API change if a new feature comes up and we start wanting to have a `FILE_NOT_FOUND` value.

Reply:
Sorry, I blindly trusted our internal documentation which says not OK "Not in VS 2017.", thanks for the correction (I updated the document).

Regardless, I agree that `bool` doesn't change anything to the generated code. My opinion is it could improve readability though, especially as a return value of functions, to make it clear that a return value of 0 means (the condition expressed by the name of the function is) false, as opposed to most functions returning `int` where 0 means success. However, I think this only makes sense if this is a convention applied consistently throughout the code base, which means having a team-wide discussion about it first, and I think we have other more urgent things to discuss/do.

So, let's stick to using `uint8_t` for now.

Reply:
Changed it to `unsigned char` - is that acceptable or would uint8_t be better?